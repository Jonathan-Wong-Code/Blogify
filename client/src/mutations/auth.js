import gql from "graphql-tag";

export const LOGOUT_MUTATION = gql`
  mutation logout {
    logout {
      message
    }
  }
`;

export const LOGIN_MUTATION = gql`
  mutation login($email: String!, $password: String!) {
    login(data: { email: $email, password: $password }) {
      token
      user {
        name
        email
        role
      }
    }
  }
`;

export const SIGNUP_MUTATION = gql`
  mutation signup(
    $name: String!
    $email: String!
    $password: String!
    $confirmPassword: String!
  ) {
    signup(
      data: {
        name: $name
        email: $email
        password: $password
        confirmPassword: $confirmPassword
      }
    ) {
      token
      user {
        name
        email
        _id
        role
      }
    }
  }
`;
