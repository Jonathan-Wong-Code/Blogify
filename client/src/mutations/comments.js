import gql from "graphql-tag";

export const CREATE_COMMENT = gql`
  mutation createComment($text: String!, $postId: ID!) {
    createComment(data: { text: $text }, postId: $postId) {
      text
      _id
      author {
        name
        _id
      }
    }
  }
`;

export const DELETE_COMMENT = gql`
  mutation deleteComment($commentId: ID!) {
    deleteComment(commentId: $commentId) {
      _id
    }
  }
`;

export const UPDATE_COMMENT = gql`
  mutation updateComment($commentId: ID!, $text: String!) {
    updateComment(commentId: $commentId, data: { text: $text }) {
      text
      _id
      author {
        name
        _id
      }
    }
  }
`;
