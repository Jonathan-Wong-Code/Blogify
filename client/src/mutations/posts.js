import gql from "graphql-tag";

export const CREATE_POST = gql`
  mutation createPost($title: String!, $body: String!) {
    createPost(data: { title: $title, body: $body }) {
      _id
      title
      body
      author {
        name
      }
    }
  }
`;

export const UPDATE_POST = gql`
  mutation updatePost($title: String!, $body: String!) {
    updatePost(data: { title: $title, body: $body }) {
      _id
      title
      body
      author {
        name
      }
    }
  }
`;

export const DELETE_POST = gql`
  mutation deleteAnyPost($id: ID!) {
    deleteAnyPost(id: $id) {
      _id
    }
  }
`;
