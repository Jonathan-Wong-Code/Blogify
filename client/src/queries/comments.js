import gql from "graphql-tag";

export const COMMENTS_BY_POST = gql`
  mutation commentsByPost($postId: ID!) {
    commentsByPost(postId: $postId) {
      text
      _id
      likes {
        _id
        name
      }
      author {
        name
        _id
      }
    }
  }
`;
