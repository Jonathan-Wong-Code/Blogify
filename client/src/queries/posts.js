import gql from "graphql-tag";
import { POST_W_AUTHOR } from "../fragments";

export const GET_ALL_POSTS = gql`
  query allPosts(
    $title: String
    $body: String
    $published: Boolean
    $sort: String
    $limit: Float
    $page: Float
  ) {
    allPosts(
      queryParams: {
        title: $title
        body: $body
        published: $published
        sort: $sort
        limit: $limit
        page: $page
      }
    ) {
      ...PostWithAuthor
    }
  }
  ${POST_W_AUTHOR}
`;

export const GET_MY_POSTS = gql`
  query myPosts(
    $title: String
    $body: String
    $published: Boolean
    $sort: String
    $limit: Float
    $page: Float
  ) {
    myPosts(
      queryParams: {
        title: $title
        body: $body
        published: $published
        sort: $sort
        limit: $limit
        page: $page
      }
    ) {
      ...PostWithAuthor
    }
  }
  ${POST_W_AUTHOR}
`;

export const GET_PUBLIC_POST = gql`
  query publicPost($id: ID!) {
    publicPost(id: $id) {
      ...PostWithAuthor
      comments {
        text
        _id
        author {
          name
          _id
        }
      }
    }
  }
  ${POST_W_AUTHOR}
`;

export const GET_PRIVATE_POST = gql`
  query privatePost($id: ID!) {
    privatePost(id: $id) {
      ...PostWithAuthor
      published
      comments {
        text
        _id
        author {
          name
          _id
        }
      }
    }
  }
  ${POST_W_AUTHOR}
`;

export const GET_NUM_PUBLIC_POSTS = gql`
  query getNumPublicPosts($title: String, $body: String) {
    getNumPublicPosts(data: { title: $title, body: $body }) {
      numPosts
    }
  }
`;
