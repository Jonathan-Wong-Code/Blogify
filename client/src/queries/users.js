import gql from "graphql-tag";

export const GET_ALL_USERS = gql`
  query users {
    users {
      name
      email
      _id
    }
  }
`;

export const GET_MY_PROFILE = gql`
  query me {
    me {
      name
      email
      _id
    }
  }
`;
