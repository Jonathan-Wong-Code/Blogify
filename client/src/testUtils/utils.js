import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { AuthProvider } from "../context/auth";
import { ApolloProvider } from "@apollo/react-hooks";

export const renderContextRouterApollo = ui => {
  return {
    ...render(
      <ApolloProvider client={{}}>
        <MemoryRouter>
          <AuthProvider>{ui}</AuthProvider>
        </MemoryRouter>
      </ApolloProvider>
    )
  };
};
