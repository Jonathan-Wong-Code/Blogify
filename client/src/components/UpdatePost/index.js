import React from "react";
import PostForm from "../PostForm";
import { useMutation, useQuery } from "@apollo/react-hooks";

import { UPDATE_POST } from "../../mutations/posts";
import { GET_PRIVATE_POST } from "../../queries/posts";

function UpdatePost({ history, match }) {
  const { id } = match.params;

  const [
    updatePost,
    { loading: updateLoading, error: updateError }
  ] = useMutation(UPDATE_POST, {
    update: (cache, { data: { updatePost } }) => {
      // We can pass variables after our query if it takes params
      cache.writeData({
        query: GET_PRIVATE_POST,
        variables: { id: updatePost._id },
        data: { privatePost: updatePost }
      });
    },

    onCompleted: ({ updatePost }) => {
      history.push(`/my-posts/${updatePost._id}`);
    }
  });

  const { error, data, loading } = useQuery(GET_PRIVATE_POST, {
    variables: { id }
  });
  if (loading) return <p data-testid="update-post-query-loading">Loading...</p>;
  if (error)
    return <p data-testid="update-post-query-error">{error.message}</p>;
  const { privatePost: editedPost } = data;
  return (
    <PostForm
      type="update"
      updatePost={updatePost}
      editedPost={editedPost}
      loading={updateLoading}
      error={updateError}
    />
  );
}

export default UpdatePost;
