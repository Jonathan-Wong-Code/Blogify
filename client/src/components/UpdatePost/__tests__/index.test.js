import React from "react";
import { MockedProvider } from "@apollo/react-testing";
import "@testing-library/jest-dom/extend-expect";
import { renderContextRouter } from "../../../testUtils/utils";
import { UPDATE_POST } from "../../../mutations/posts";
import { GET_PRIVATE_POST } from "../../../queries/posts";
import { Route, MemoryRouter } from "react-router-dom";
import { render, fireEvent, wait, cleanup, act } from "@testing-library/react";

import UpdatePost from "..";
import { GraphQLError } from "graphql";

const post = {
  title: "testTitle",
  body: "Testing body",
  _id: "test1234",
  __typename: "Post",
  author: {
    __typename: "User",
    name: "test"
  }
};

let updateMutationCalled;

const updateSuccess = [
  {
    request: {
      query: GET_PRIVATE_POST,
      variables: {
        id: post._id
      }
    },

    result: {
      data: {
        privatePost: post
      }
    }
  },
  {
    request: {
      query: UPDATE_POST,
      variables: {
        title: "testTitle",
        body: "Testing body"
      }
    },

    result: () => {
      updateMutationCalled = true;
      return { data: { updatePost: post } };
    }
  }
];

const match = { params: { id: "test1234" } };

beforeEach(() => {
  jest.clearAllMocks();
  updateMutationCalled = false;
});

afterEach(cleanup);

describe("<UpdatePost>", () => {
  it("renders", async () => {
    const { container } = renderContextRouter(
      <MockedProvider mocks={updateSuccess} addTypename={false}>
        <UpdatePost match={match} />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();

    await wait(() => {
      expect(container).toMatchSnapshot();
    });
  });
});
