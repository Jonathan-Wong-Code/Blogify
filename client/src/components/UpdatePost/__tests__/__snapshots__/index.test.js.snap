// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UpdatePost> renders 1`] = `
<div>
  <p>
    Loading...
  </p>
</div>
`;

exports[`<UpdatePost> renders 2`] = `
<div>
  <section>
    <form
      action=""
    >
      <label
        for="post-title"
      >
        Title:
      </label>
      <input
        id="post-title"
        type="text"
        value="testTitle"
      />
      <label
        for="post-body"
      >
        Body:
      </label>
      <textarea
        cols="50"
        id="post-body"
        name="post-body"
        rows="20"
      >
        Testing body
      </textarea>
      <button>
        Submit
      </button>
    </form>
  </section>
</div>
`;
